Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Utilities/MaterialInstancing.cs

```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace Microsoft.MixedReality.Toolkit.UI
+{
+    public class MaterialInstancing : MonoBehaviour
+    {
+        private void Awake()
+        {
+            var images =  gameObject.GetComponentsInChildren<Image>(true);
+            foreach (var image in images)
+            {
+                image.material = Instantiate(image.material);
+            }
```

Comment:
Performance question here: do we really need to do this for every child image/graphic?  Can we instead do something smart to dynamically create one when we need to set a property on it, then revert back when/if the properties go back to the default?

Reply:
> Performance question here: do we really need to do this for every child image/graphic? Can we instead do something smart to dynamically create one when we need to set a property on it, then revert back when/if the properties go back to the default?

I am not a huge fan of this either, but I have not been able to come up with a better approach.  If there was a single script that was modifying material properties it would have been achievable to do what you are recommending. 
We could potentially write some sort of Material extension system (we used to have it in some old codebases) which does ref counting and achieves what you are recommending above. In fact, it could also be written to perform batched operations per frame to boost up performance. It would be a feature in itself though and would need some thinking, with the added expectation from consumers to use the MaterialExetnsion instead of material for perf improvements.

Reply:
@Troy-Ferrell  - do you have any thoughts around this ? Are there any shader experts in MRTK team who would have thoughts around this?

Reply:
It would probably be better to have a wrapper static class that surfaces get/set color/float shader properties. This class would handle instantiating a new material instance and then set values. At least then instantiation would be lazy loaded on only the things that change. You could also just keep a set of graphic references for whether a new material needs to be created or not. Then have the UGUIShader theme class utilize this

Reply:
So i refactored the  MaterialInstancing class  and hooked that up into unity ui targetted shader themes. I also removed the Color property changes fromt he existing InteractableColorTheme to a separate one that works for Graphic object. How does this feel ?